exampleCourse/serverFilesCourse/r_autograder/pltest.R

```diff
@@ -0,0 +1,157 @@
+## Script Requirements ----
+
+# Note: Packages listed here are found on the PrairieLearn R docker image
+
+# Use testthat to test the student code
+library("testthat")
+
+# Use stringr to simplify string manipulation
+library("stringr")
+
+# Use jsonlite to export the test results under a PrairieLearn format
+library("jsonlite")
+
+## Script hierarchy
+# /grade
+# |- data/
+# |- results/
+# |- run/                # <--- Working directory
+# |- serverFilesCourse/
+# |- shared/
+# |- student/
+# |- tests/
+
+## Run Unit Tests ----
+
+# Automatically test the directory
+# This helps avoid needing to build a package on a per question basis.
+# To understand testing outside of a package, it's mainly documented in an
+# issue ticket: https://github.com/r-lib/testthat/issues/659
+
+# Begin the process of grading each student's code
+# Add a capture statement to avoid allowing students to output
+# content to console.
+log <- capture.output({
+    test_data_as_list <- testthat::test_dir("../tests",
+                                           reporter = testthat::ListReporter)
+})
+
+# Convert from list to a data.frame
+test_data = as.data.frame(test_data_as_list)
+
+## Compute Unit Test Correctness ----
+
+# Retrieve question totals from the test description. To facilitate this,
+# we assume the format is given as...
+#
+# NumericPoints: test description
+#
+# To ensure no issue, we remove all none-numeric values on the left and right
+# of the colon separator.
+
+# Obtain the maximum points
+test_data$max_points = stringr::str_replace(
+    test_data$test,
+    pattern = ".*([:digit:]+).*:.*",
+    replacement = "\\1"
+)
+
+# Convert retrieved maximum points to a numeric value
+test_data$max_points = as.numeric(test_data$max_points)
+
+# Calculate the number of points the student received
+test_data$earned_points = ifelse(test_data$passed >= 1,
+                                 test_data$max_points,
+                                 0.0)
+
+# Now, let's aim to extract the description values
+test_data$description = stringr::str_replace(
+    test_data$test,
+    pattern = ".*[:digit:]+.*:[:space:](.*)",
+    replacement = "\\1"
+)
+
+## Format the Results to Match External Grading Format ----
+
+# Format the results of each inline test.
+pl_test_list_format = function(i, x) {
+
+    pl_message_statement = function(error, fail) {
+        if(isTRUE(error)) {
+            "Your code encounted an error. Please verify it works."
```

Comment:
You don’t show the student the error? I feel like errors should passed back to the student so they can fix them?

Reply:
I wasn't sure how indepth the error was given the entry on readthedocs. I also wanted to provide a means to avoid having this be viewed as a live code editor.

Reply:
I think it’s better to err on the side of giving students too much information. While it’s suboptimal if they do live debugging in the editor, I think this is better than them getting frustrated when they can’t pass a test, and they also can’t see what’s wrong. For example, I’ve seen students with working code in a local dev environment, who have a copy-paste problem (e.g., utf8 issues when pasted from Windows) and would be really unable to debug it if they can’t see detailed information about the return values and errors from their code when run in the grader.

I guess we should document this somewhere? 😄